ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/transaction/TransactionSignatureService.kt

```diff
@@ -0,0 +1,57 @@
+package net.corda.v5.ledger.common.transaction
+
+import net.corda.v5.application.crypto.DigitalSignatureAndMetadata
+import net.corda.v5.application.crypto.DigitalSignatureVerificationService
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.crypto.merkle.MerkleProof
+import java.security.PublicKey
+
+/**
+ * TransactionSignatureService can be used to sign and verify transaction signatures.
+ * It supports both single and batch signatures.
+ */
+
+interface TransactionSignatureService {
+    /**
+     * Signs a transaction id with all the available keys.
+     *
+     * @param transactionId The transaction id to be signed.
+     * @param publicKeys The keys which should be tried to sign with.
+     *
+     * @returns Resulting signatures.
+     *
+     * @throws TransactionNoAvailableKeysException If none of the keys are available.
+     */
+    @Suspendable
+    fun sign(transactionId: SecureHash, publicKeys: Iterable<PublicKey>): List<DigitalSignatureAndMetadata>
```

Comment:
Is there a reason we are passing in just the transaction id here, rather than the full transaction as we do with the overload? Wouldn't this be more future-proof and consistent?

Reply:
Originally I started with that approach. Then realized that metadata is not used at all, and `TransactionWithMetadata` just limits its usability. Also that time we were not too sure if ledger will be the final layer.
But this last argument is already clarified and I agree that it would be more consistent with `TransactionWithMetadata`.
So I changed it back to `TransactionWithMetadata`.